client/me/purchases/index.js

```diff
@@ -58,11 +58,14 @@ export default function( router ) {
 		clientRender
 	);
 
+	/**
+	 * The siteSelection middleware has been removed from this root.
+	 * No selected site!
+	 */
```

Comment:
`s/root/route/` 🤦‍♂️ 

---

The goal of this PR is to allow purchasers of Jetpack plans on sites that are disconnected to manage and remove the purchase.

A disconnected Jetpack site is removed from the users sites, we do not have access to the site object.

The `siteSelection` middleware (below) does a bunch of things. Primarily, it looks for a site slug or site ID in the route, tries to find that in the users sites, and loads sets a bit of state indicating the selected site (`state.ui.selectedSiteId`). Information about this site is then accessible via a bunch of selected site selectors.

https://github.com/Automattic/wp-calypso/blob/e742318d277ee549fe7d851d26b991490e8e9588/client/my-sites/controller.js#L300-L415

This works well around a lot of Calypso where we actually have sites. The purchases were designed with this same pattern in mind. However, there's no real constraint that the user must have a site and a purchase. This Jetpack case illustrates that a user may have a purchase without having site data available.

This means that rather than relying on this magical (`siteSelection` + selected site selectors) site data, we should manually select a site (we have enough data in the purchase object) and avoid relying on the site data when possible as it may not be present.

I imagine a future where we remove site slugs from the purchase routes (they're irrelevant, purchase ID is enough to get site data) and eliminate dependence on the site object, but that requires a lot more work. I've been chipping away at it for a while, but this PR goes as far as is necessary for the immediate goal of allowing disconnected Jetpack site purchases to be managed.

---

Do you think we need a more descriptive comment here?

Reply:
Just to clarify, I highlighted this case with a comment because the other sections continue to include the `siteSelection` middleware.

In many cases, there will continue to be _some_ selected site, we just have no guarantees that it will be the site we expect and we shouldn't be using those selectors.

This has gotten me thinking, however. I wonder if it makes sense to explicitly remove any selected site when we're in this route:

https://github.com/Automattic/wp-calypso/blob/e742318d277ee549fe7d851d26b991490e8e9588/client/state/ui/actions.js#L31-L42

Reply:
Nah, calling `setAllSitesSelected()` here is just as un-semantic and confusing IMO. If anything, we could run `setSelectedSite()` with the `siteId` fetched from `purchase`. But doing it that way also opens up the door for confusion (can I rely on `getSelectedSiteId()` or not? Should I use `purchase.siteId` instead)?

My preference is for just killing the beast -- i.e. the site slug route fragment -- altogether in a follow-up PR.

Reply:
It gets pretty involved to remove the site slug from any route. It pretty much necessitates removing it from all routes across the section because you need it in all routes just for navigation.

I do agree that's the end goal 👍